### PR TITLE
Eventstream bug fixes

### PR DESCRIPTION
**Pull request by wesboyt**
*Issue #, if available:*

*Description of changes:*
Fixes List, Sexp, and Struct so that they do not unnecesarily recurse.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/IonEvent.ts [hunk 1]
```diff
@@ -404,12 +404,17 @@ class IonStructEvent extends AbsIonContainerEvent {//no embed support as of yet.
...
-                    paired[j] = matchFound;
+                    if (matchFound && actualEvents[i].eventType === IonEventType.CONTAINER_START) {
+                        for (let k = j; k < expectedEvents[j].ionValue.length; k++) {
+                            paired[k] = true;
+                        }
+                        i += actualEvents[i].ionValue.length;
+                    }
                 }
             }
         }
```
Comment (pbcornell) [inline src/IonEvent.ts]: if `expectedEvents` matches all the events of `actualEvents`, but `expectedEvents` contains additional, unmatched events, i believe this method will incorrectly return `true`

File: src/IonEvent.ts [hunk 2]
```diff
@@ -404,12 +404,17 @@ class IonStructEvent extends AbsIonContainerEvent {//no embed support as of yet.
...
-                if(!paired[j]) {
+            for (let j : number = 0; !matchFound && j < expectedEvents.length; j++) {
+                if (!paired[j]) {
                     matchFound = actualEvents[i].equals(expectedEvents[j]);
-                    paired[j] = matchFound;
+                    if (matchFound && actualEvents[i].eventType === IonEventType.CONTAINER_START) {
+                        for (let k = j; k < expectedEvents[j].ionValue.length; k++) {
+                            paired[k] = true;
+                        }
+                        i += actualEvents[i].ionValue.length;
```
Comment (pbcornell) [inline src/IonEvent.ts]: will this cause some events to go unchecked?  i haven't been able to show that this is a problem, but my concern is similar to that raised below for lists/sexps

File: src/IonEvent.ts (hunk 1 above)
Comment (wesboyt) [reply src/IonEvent.ts]: this.structsEqual(this.ionValue, expected.ionValue) && this.structsEqual(expected.ionValue, this.ionValue);

File: src/IonEvent.ts [hunk 3]
```diff
@@ -429,6 +434,8 @@ class IonListEvent extends AbsIonContainerEvent {
         for(let i : number = 0; i < this.ionValue.length; i++){
             if(!this.ionValue[i].equals(expected.ionValue[i])){
                 return false;
+            } else if(this.ionValue[i].eventType === IonEventType.CONTAINER_START) {
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: dear god

File: src/IonEvent.ts (hunk 2 above)
Comment (wesboyt) [reply src/IonEvent.ts]: the issue was in lists and sexps where the .length call was returning undefined and the += was translating i to NaN. It is now fixed.

File: src/IonEvent.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: fixed

File: src/IonEvent.ts [hunk 4]
```diff
@@ -390,7 +390,7 @@ abstract class AbsIonContainerEvent extends AbstractIonEvent{
 }
 
 class IonStructEvent extends AbsIonContainerEvent {//no embed support as of yet.
-    constructor(eventType : IonEventType, ionType : IonType, fieldName : string, annotations : string[], depth : number) {
+     constructor(eventType : IonEventType, ionType : IonType, fieldName : string, annotations : string[], depth : number) {
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: done

File: src/IonEvent.ts (hunk 1 above)
Comment (wesboyt) [reply src/IonEvent.ts]: > * please try removing good/equivs/lists.ion and good/equivs/sexps.ion from the skipList and see if TravisCI no longer timesout (see the toSkipList() method in iontests.js)
> * for the value `{a: 1, b: {c: 3}, d: 4}`, event stream has the following structure:
> 
> ```
>   SCALAR 1 int a: 1
>   CONTAINER_START 1 struct b: 
>     SCALAR 2 int c: 3
>     CONTAINER_END 1
>   SCALAR 2 int c: 3   // should the event for nested field 'c' appear here as well as above?
>   CONTAINER_END 1
>   SCALAR 1 int d: 4
>   CONTAINER_END 0
> ```
> 
> is this the expected structure?

a container start event indexes all events within (not itself) as well as its container end event.

Comment (wesboyt) [reply src/IonEvent.ts]: I removed the test vectors from the skip list and the build is not timing out.

File: src/IonEvent.ts [hunk 5]
```diff
@@ -445,6 +452,8 @@ class IonSexpEvent extends AbsIonContainerEvent {
         for(let i : number = 0; i < this.ionValue.length; i++){
             if(!this.ionValue[i].equals(expected.ionValue[i])){
                 return false;
+            } else if(this.ionValue[i].eventType === IonEventType.CONTAINER_START) {
+                i += this.ionValue[i].ionValue.length;
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: IonValue isnt a type?

File: src/IonEvent.ts [hunk 6]
```diff
@@ -429,6 +434,8 @@ class IonListEvent extends AbsIonContainerEvent {
...
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: You can open an issue for it if you'd like, i don't think altering it adds very much value.

File: src/IonEvent.ts [hunk 7]
```diff
@@ -404,12 +404,17 @@ class IonStructEvent extends AbsIonContainerEvent {//no embed support as of yet.
     structsEqual(actualEvents : AbstractIonEvent[], expectedEvents : AbstractIonEvent[]) : boolean {
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply src/IonEvent.ts]: yes structs can be equal when out of order.

File: src/IonEvent.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply src/IonEvent.ts]: Perhaps it's not helpful to you as the original author, but for a new contributor code like:

```typescript
i += this.ionValue[i].ionValue.length;
```

is not self-explanatory, particularly in a weakly typed language.

File: src/IonEvent.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply src/IonEvent.ts]: It is true that `ion-js` does not have a formal `IonValue` type as (e.g.) IonJava does. However, I would expect a field called `ionValue` to contain a value read from Ion, not an `IonEvent`.

File: src/IonEvent.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply src/IonEvent.ts]: Sorry, allow me to rephrase: I'm surprised that this requires a triply-nested `for` loop. Is there not a more efficient way to evaluate their equality, perhaps by using a different data structure?

**Comment by almann** [comment]
Can you please rebase this against `master`, thanks!

File: src/IonEvent.ts (hunk 1 above)
Comment (wesboyt) [reply src/IonEvent.ts]: // different field/value pairs
[
    { a:4, a:4, b:6 },
    { a:4, b:6, b:6 },
]

**Pull request closed without merging**